docx/oxml/numbering.py

```diff
@@ -152,30 +152,93 @@ def get_num_for_p(self, p, styles_cache):
         """
         Returns list item for the given paragraph.
         """
-        numPr = p.pPr.get_numPr(styles_cache)
-        ilvl, numId = numPr.ilvl, numPr.numId.val
-        ilvl = ilvl.val if ilvl is not None else 0
+
+        def get_ilvl_and_numId(paragraph):
+            """
+            Return ``ilvl`` and ``numId`` for the given ``paragraph``.
+            """
+            para_numPr = paragraph.pPr.get_numPr(styles_cache)
+            para_ilvl, para_numId = para_numPr.ilvl, para_numPr.numId.val
+            para_ilvl = para_ilvl.val if para_ilvl is not None else 0
+            return para_ilvl, para_numId
+
+        def get_preceding_paragraphs_numIds(p, p_ilvl, p_numId):
+            """
+            Return preceding siblings ``numId`` that are in the same numbered list as the paragraph ``p``.
+            Paragraphs are in the same list if they are on the same level (``p_ilvl``), and either
+            have the same ``p_numId`` or have the same ``pStyle.val``.
+            Skips unnumbered paragraphs within the numbering list.
+            Stops on the paragraph that is on the lower level.
+            """
+            pStyle = p.pPr.pStyle
+            for para in p.itersiblings(preceding=True):
```

Comment:
would it make sense to rename `para` to something like `pp` or `prev_p` which would associate with any previous paragraphs